rpc/src/module/subscription.rs

```diff
@@ -0,0 +1,170 @@
+use ckb_logger::error;
+use ckb_notify::NotifyController;
+use crossbeam_channel::select;
+use jsonrpc_core::{futures::Future, Metadata, Result};
+use jsonrpc_derive::rpc;
+use jsonrpc_pubsub::{
+    typed::{Sink, Subscriber},
+    PubSubMetadata, Session, SubscriptionId,
+};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::sync::Arc;
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    RwLock,
+};
+use std::thread;
+
+#[derive(Clone, Debug)]
+pub struct SubscriptionSession {
+    pub(crate) subscription_ids: Arc<RwLock<HashSet<SubscriptionId>>>,
+    pub(crate) session: Arc<Session>,
+}
+
+impl SubscriptionSession {
+    pub fn new(session: Session) -> Self {
+        Self {
+            subscription_ids: Arc::new(RwLock::new(HashSet::new())),
+            session: Arc::new(session),
+        }
+    }
+}
+
+impl Metadata for SubscriptionSession {}
+
+impl PubSubMetadata for SubscriptionSession {
+    fn session(&self) -> Option<Arc<Session>> {
+        Some(Arc::clone(&self.session))
+    }
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Eq, Hash)]
+#[serde(rename_all = "snake_case")]
+pub enum Topic {
+    NewTipHeader,
+    NewTipBlock,
+}
+
+#[allow(clippy::needless_return)]
+#[rpc(server)]
+pub trait SubscriptionRpc {
+    type Metadata;
+
+    #[pubsub(subscription = "subscribe", subscribe, name = "subscribe")]
+    fn subscribe(&self, meta: Self::Metadata, subscriber: Subscriber<String>, topic: Topic);
+
+    #[pubsub(subscription = "subscribe", unsubscribe, name = "unsubscribe")]
+    fn unsubscribe(&self, meta: Option<Self::Metadata>, id: SubscriptionId) -> Result<bool>;
+}
+
+type Subscribers = HashMap<SubscriptionId, Sink<String>>;
+
+#[derive(Default)]
+pub struct SubscriptionRpcImpl {
+    pub(crate) id_generator: AtomicUsize,
+    pub(crate) subscribers: Arc<RwLock<HashMap<Topic, Subscribers>>>,
+}
+
+impl SubscriptionRpc for SubscriptionRpcImpl {
+    type Metadata = Option<SubscriptionSession>;
+
+    fn subscribe(&self, meta: Self::Metadata, subscriber: Subscriber<String>, topic: Topic) {
+        if let Some(session) = meta {
+            let id =
+                SubscriptionId::Number(self.id_generator.fetch_add(1, Ordering::SeqCst) as u64);
+            if let Ok(sink) = subscriber.assign_id(id.clone()) {
+                let mut subscribers = self
+                    .subscribers
+                    .write()
+                    .expect("acquiring subscribers write lock");
+                subscribers
+                    .entry(topic)
+                    .or_default()
+                    .insert(id.clone(), sink);
+
+                session
+                    .subscription_ids
+                    .write()
+                    .expect("acquiring subscription_ids write lock")
+                    .insert(id);
+            }
+        }
+    }
+
+    fn unsubscribe(&self, meta: Option<Self::Metadata>, id: SubscriptionId) -> Result<bool> {
```

Comment:
What about using `id: Uint64`?

Reply:
Updated subscription id to hex string, please review again, thanks.